src/coreclr/jit/morph.cpp

```diff
@@ -5183,6 +5183,12 @@ GenTree* Compiler::fgMorphArrayIndex(GenTree* tree)
 
     noway_assert(elemTyp != TYP_STRUCT || elemStructType != nullptr);
 
+    if (opts.OptimizationEnabled())
+    {
+        // Fold possible constant expressions
+        asIndex->Index() = gtFoldExpr(asIndex->Index());
```

Comment:
"hello"[1 + 1] wasn't folded before this change (it's needed for `b[b.Length - 1]` etc)